### DISCUSSION TITLE
Pool's overflow connections not disconnecting as expected

### DISCUSSION THREAD

**Discussion by gricey432** [sqlalchemy/sqlalchemy#6638] [Usage Questions] [answered]
**Describe the bug**
The `max_overflow` is [documented](https://docs.sqlalchemy.org/en/14/core/pooling.html#sqlalchemy.pool.QueuePool.params.max_overflow) as

> When those additional connections are returned to the pool, they are disconnected and discarded.

However I'm seeing connections held open after calling `close()` on them.

**To Reproduce**
```py
import sqlalchemy as sa
from sqlalchemy.orm import sessionmaker

engine = sa.create_engine(
    'postgresql://webapp:webapp@localhost:5432/test',
    pool_size=0,
    echo=True,
    echo_pool='debug',
)
session_factory = sessionmaker(bind=engine)

sess = session_factory()
print(engine.pool.status())
# Pool size: 0  Connections in pool: 0 Current Overflow: 0 Current Checked out connections: 0

sess.execute("SELECT 1").fetchall()
print(engine.pool.status())
# Pool size: 0  Connections in pool: 0 Current Overflow: 1 Current Checked out connections: 1

sess.close()

print(engine.pool.status())
# Breakpoint here and checking the postgres server shows one idle connection
# Pool size: 0  Connections in pool: 1 Current Overflow: 1 Current Checked out connections: 0
```

**Error**
```
Pool size: 0  Connections in pool: 0 Current Overflow: 0 Current Checked out connections: 0
2021-06-15 13:21:23,963 DEBUG sqlalchemy.pool.impl.QueuePool Created new connection <connection object at 0x0000025C52D98BF8; dsn: 'user=webapp password=xxx dbname=test host=localhost port=5432', closed: 0>
2021-06-15 13:21:23,972 INFO sqlalchemy.engine.Engine select version()
2021-06-15 13:21:23,972 INFO sqlalchemy.engine.Engine [raw sql] {}
2021-06-15 13:21:23,975 INFO sqlalchemy.engine.Engine select current_schema()
2021-06-15 13:21:23,975 INFO sqlalchemy.engine.Engine [raw sql] {}
2021-06-15 13:21:23,977 INFO sqlalchemy.engine.Engine show standard_conforming_strings
2021-06-15 13:21:23,978 INFO sqlalchemy.engine.Engine [raw sql] {}
2021-06-15 13:21:23,980 DEBUG sqlalchemy.pool.impl.QueuePool Connection <connection object at 0x0000025C52D98BF8; dsn: 'user=webapp password=xxx dbname=test host=localhost port=5432', closed: 0> checked out from pool
2021-06-15 13:21:23,980 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2021-06-15 13:21:23,981 INFO sqlalchemy.engine.Engine SELECT 1
2021-06-15 13:21:23,981 INFO sqlalchemy.engine.Engine [generated in 0.00016s] {}
Pool size: 0  Connections in pool: 0 Current Overflow: 1 Current Checked out connections: 1
2021-06-15 13:21:23,983 INFO sqlalchemy.engine.Engine ROLLBACK
2021-06-15 13:21:23,988 DEBUG sqlalchemy.pool.impl.QueuePool Connection <connection object at 0x0000025C52D98BF8; dsn: 'user=webapp password=xxx dbname=test host=localhost port=5432', closed: 0> being returned to pool
2021-06-15 13:21:23,988 DEBUG sqlalchemy.pool.impl.QueuePool Connection <connection object at 0x0000025C52D98BF8; dsn: 'user=webapp password=xxx dbname=test host=localhost port=5432', closed: 0> rollback-on-return
Pool size: 0  Connections in pool: 1 Current Overflow: 1 Current Checked out connections: 0
```

**Versions.**
 - OS: Windows 10
 - Python: 3.6.8
 - SQLAlchemy: 1.3.24 & 1.4.18
 - Database: postgresql 11
 - DBAPI: psycopg2

**Have a nice day!**
You too

**Comment by zzzeek** [comment, marked as answer]
hey there -

SQLAlchemy's `QueuePool` is based on the Python [queue.Queue](https://docs.python.org/3/library/queue.html) object.     One particular oddness of this object is that [setting the size of the Queue to zero indicates unlimited storage](https://docs.python.org/3/library/queue.html#queue.Queue):

>  If maxsize is less than or equal to zero, the queue size is infinite.

SQLAlchemy's `QueuePool` therefore documents this as well in the parameter string for [pool_size](https://docs.sqlalchemy.org/en/14/core/engines.html?highlight=pool_size#sqlalchemy.create_engine.params.pool_size):

> With QueuePool, a pool_size setting of 0 indicates no limit; to disable pooling, set poolclass to NullPool instead.

Therefore if you are looking for a pool that does not hold onto any connections after use, you should use NullPool as [demonstrated in this example](https://docs.sqlalchemy.org/en/14/core/pooling.html?highlight=nullpool#switching-pool-implementations).

hope this helps!

**Reply by gricey432** [reply]
Eurgh, I was so focused in on the overflow that I didn't even notice that. Really appreciate the help!

**Reply by gricey432** [reply]
This leads to a bit of a strange situation, if I were to say

> I want an infinitely big QueuePool and I want all the connections to be kept open (no overflow)

```
engine = sa.create_engine(
    'postgresql://webapp:webapp@localhost:5432/test',
    pool_size=0,
    max_overflow=0,
    echo=True,
    echo_pool='debug',
)
session_factory = sessionmaker(bind=engine)
sess = session_factory()
print(engine.pool.status())
sess.execute("SELECT 1").fetchall()
```

The above should have an unlimited pool size from the `pool_size=0`, but the [implementation](https://github.com/sqlalchemy/sqlalchemy/blob/7f29593823bd80473fd53264942da6631399f814/lib/sqlalchemy/pool/impl.py#L123) uses the `overflow` to achieve infinite connections. So the above code never gets a connection and errors.

```
sqlalchemy.exc.TimeoutError: QueuePool limit of size 0 overflow 0 reached, connection timed out, timeout 30.00 (Background on this error at: http://sqlalche.me/e/14/3o7r)
```

**Reply by zzzeek** [reply]
yeah zero should likely just be disallowed.  it's just something that slipped through.

**Reply by gricey432** [reply]
For now while it is still an option, would you accept a PR changing the 
```
Pool size: 0  Connections in pool: 1 Current Overflow: 1 Current Checked out connections: 0
```
To say
```
Pool size: infinite  Connections in pool: 1 Current Overflow: 1 Current Checked out connections: 0
```
To improve clarity on this when debugging?
Not sure if that would count as a breaking change

**Reply by gricey432** [reply]
It looks like there's no way to achieve the opposite too; overflow only pool with a connection cap.

If I use `QueuePool` I can't set the `pool_size` to 0 otherwise I'll get unlimited persistent connections.
If I use `NullPool` I can't set a limit on the number of connections.

I've made this custom pool to solve this scenario.

```
class CappedNullPool(NullPool):
    """A Pool which does not pool connections and has a cap on the
      maximum number of connections
    """
    def __init__(
        self,
        creator,
        pool_size=0,
        **kw
    ):
        """
        :param pool_size: The maximum overflow size of the pool.
          ``pool_size`` can be set to 0 to indicate no size limit.
          Defaults to 0 (no limit).

        :param \**kw: Other keyword arguments including
          :paramref:`_pool.Pool.recycle`, :paramref:`_pool.Pool.echo`,
          :paramref:`_pool.Pool.reset_on_return` and others are passed to the
          :class:`_pool.Pool` constructor.
        """
        NullPool.__init__(self, creator, **kw)
        self._pool_size = pool_size
        self._checkedout = 0
        self._checkout_lock = threading.Lock()

    def _do_get(self):
        with self._checkout_lock:
            if self._pool_size == 0 or self._checkedout < self._pool_size:
                self._checkedout += 1
            else:
                raise ValueError(f"CappedNullPool limit of {self._pool_size} reached")
        return super(CappedNullPool, self)._do_get()

    def _do_return_conn(self, conn):
        with self._checkout_lock:
            self._checkedout -= 1
        super(CappedNullPool, self)._do_return_conn(conn)

    def status(self):
        return f"Pool size: {self.size()} Current Checked out connections: {self.checkedout()}"

    def size(self):
        return self._pool_size

    def checkedout(self):
        return self._checkedout
```

Not sure if it would be useful to the library / anyone else. But also if you instead go ahead with removing the `0` option from `QueuePool` then it may become moot.

🙂

**Reply by zzzeek** [reply]
I would first want the size of "0" to emit a deprecation warning.   then if you'd like, the message can say something like "pool size: 0 (queue is unlimited) ..."
